### PR TITLE
[datastore] Mongodb source & target (system test also) 

### PR DESCRIPTION
**Pull request by davesh0812**
Implement of MongoDB Source and Target for a flow.  Now there is an option to preform fs pipeline with those source & target and to get online features as well.
Also I fixed a bug - When using "as" in the feature list of fv and try to do vector.service.get({}, as_list=True), the list isn't contain the feature that we tried to change it name.

### CODE REVIEW

File: mlrun/datastore/sources.py [hunk 1]
```diff
@@ -826,6 +829,139 @@ def add_nuclio_trigger(self, function):
...
+            key_field=key_field,
+            time_field=time_field,
+            schedule=schedule,
+            start_time=start_time,
+            end_time=end_time,
+        )
+
+    def to_dataframe(self):
+        class MongoDBiter:
+            def __init__(self, collection, iter_chunksize, iter_query):
```
Comment (guy1992l) [inline mlrun/datastore/sources.py:909]: please add type hints and doc strings to all the methods and variables you added.

File: mlrun/datastore/targets.py [hunk 2]
```diff
@@ -1429,6 +1431,215 @@ def as_df(
...
+            max_events=max_events,
+            flush_after_seconds=flush_after_seconds,
+            storage_options=storage_options,
+            after_state=after_state,
+        )
+
+    def add_writer_state(
+        self, graph, after, features, key_columns=None, timestamp_key=None
+    ):
+        warnings.warn(
```
Comment (guy1992l) [inline mlrun/datastore/targets.py:1530]: we are far from 0.9.0 and if you are adding new code, I think it should be gone from the first place.

File: mlrun/datastore/targets.py [hunk 3]
```diff
@@ -1429,6 +1431,215 @@ def as_df(
         return self._df
 
 
+class MongoDBTarget(BaseStoreTarget):
```
Comment (guy1992l) [inline mlrun/datastore/targets.py:1434]: same as the source, move to the init doc string.

File: mlrun/datastore/targets.py [hunk 4]
```diff
@@ -1429,6 +1431,215 @@ def as_df(
...
+    def write_dataframe(
+        self, df, key_column=None, timestamp_key=None, chunk_id=0, **kwargs
+    ):
+        if hasattr(df, "rdd"):
+            raise ValueError("Spark is not supported")
+        else:
+            from pymongo import MongoClient
+
+            connection_string, db_name, collection_string = self._parse_url()
+            mongodb_client = MongoClient(connection_string)
```
Comment (guy1992l) [inline mlrun/datastore/targets.py:1628]: shouldn't you keep the client or you are writing just one time? Also, shouldn't you close the client (I'm just making sure you checked these stuff).

File: mlrun/feature_store/api.py [hunk 5]
```diff
@@ -265,6 +265,12 @@ def get_online_feature_service(
 
     # todo: support remote service (using remote nuclio/mlrun function if run_config)
 
+    for old_name in service.vector.get_feature_aliases().keys():
```
Comment (guy1992l) [inline mlrun/feature_store/api.py:268]: what is this?

File: mlrun/feature_store/feature_vector.py [hunk 6]
```diff
@@ -502,6 +502,11 @@ def get(self, entity_rows: List[Union[dict, list]], as_list=False):
             if not data:
                 data = None
             else:
+
+                requested_columns = [
```
Comment (guy1992l) [inline mlrun/feature_store/feature_vector.py:506]: how is that related to mongo db?

File: requirements.txt [hunk 7]
```diff
@@ -54,3 +54,4 @@ deepdiff~=5.0
 pymysql~=1.0
 inflection~=0.5.0
 python-dotenv~=0.17.0
+pymongo~=4.1.1
```
Comment (guy1992l) [inline requirements.txt:57]: I don't understand why it should be added to everyone, in my eyes its like I would put xgboost, lightgbm, catboost, tensorflow and pytorch in the requirements.

File: tests/system/feature_store/test_mongoDB.py [hunk 8]
```diff
@@ -0,0 +1,304 @@
+import datetime
```
Comment (guy1992l) [inline tests/system/feature_store/test_mongoDB.py:1]: files names shouldn't be with capital letters, rename to `test_mongo_db.py`

**Review by guy1992l — CHANGES_REQUESTED** [review]
nice job dealing with the mango ;)

* Some little chanegs are required.
* There is a design issue regarding the files (IMO), the sources.py should be a directory and in it all the sources in separate files so the requirements of MLRun which is already too long won't be longer and each user will have only what he needs.
* And, last but by no means least! Please add doc strings, type hints and comments `#` to section out your code. It is kinda painful to review long methods without it and when someone will be back to this code it will be MUCH more easier to understand.

File: requirements.txt (hunk 7 above)
Comment (guy1992l) [reply requirements.txt:57]: also, while we are at it, you did used `bason` or something like that as well no? so why no to add this as well?

File: mlrun/datastore/sources.py [hunk 9]
```diff
@@ -18,10 +18,13 @@
 from datetime import datetime
 from typing import Dict, List, Optional, Union
 
+import bson
+import pandas as pd
 import v3io
 import v3io.dataplane
 from nuclio import KafkaTrigger
 from nuclio.config import split_path
+from pymongo import MongoClient
```
[the remark this answers is not in the document]
Comment (davesh0812) [reply mlrun/datastore/sources.py:27]: I put it in inner scope like BigQuerySource

File: mlrun/feature_store/api.py (hunk 5 above)
Comment (davesh0812) [reply mlrun/feature_store/api.py:268]: @guy1992l it is not related but it is a bug I fixed  - [fix bug when using "as" if the feature list]

File: mlrun/feature_store/feature_vector.py (hunk 6 above)
Comment (davesh0812) [reply mlrun/feature_store/feature_vector.py:506]: @guy1992l it is not related but it is a bug I fixed  - [fix bug when using "as" if the feature list]

File: requirements.txt (hunk 7 above)
Comment (davesh0812) [reply requirements.txt:57]: So where you think I should write this requirements? @guy1992l

File: mlrun/datastore/sources.py [hunk 10]
```diff
@@ -875,21 +873,45 @@ def __init__(
...
+        :param schedule: string to configure scheduling of the ingestion job. For example '*/30 * * * *' will
+             cause the job to run every 30 minutes
+        :param db_name: the name of the database to access
+        :param connection_string: your mongodb connection string
+        :param collection_name: the name of the collection to access,
+                                        from the current database
+        :param spark_options: additional spark read options
+        """
+
+        MONGO_CONNECTION_STRING_ENV_VAR = "MONGO_CONNECTION_STRING"
```
Comment (guy1992l) [inline mlrun/datastore/sources.py:901]: move to be at the class level, all caps is marking a static variable, if its usage is only internal of the class then it should be at thee class level, if its only for the developers use, it should start with a `_` to be marked as protected.

File: mlrun/datastore/sources.py [hunk 11]
```diff
@@ -875,21 +873,45 @@ def __init__(
...
+        :param key_field: the column to be used as the key for events. Can be a list of keys.
+        :param time_field: the column to be parsed as the timestamp for events. Defaults to None
+        :param start_time: filters out data before this time
+        :param end_time: filters out data after this time
+        :param schedule: string to configure scheduling of the ingestion job. For example '*/30 * * * *' will
+             cause the job to run every 30 minutes
+        :param db_name: the name of the database to access
+        :param connection_string: your mongodb connection string
+        :param collection_name: the name of the collection to access,
+                                        from the current database
```
Comment (guy1992l) [inline mlrun/datastore/sources.py:897]: please align the docs :+1:

File: mlrun/datastore/targets.py (hunk 2 above)
Comment (davesh0812) [reply mlrun/datastore/targets.py:1536]: So to delete it? 
I had because all the sources have this method

File: mlrun/datastore/sources.py [hunk 12]
```diff
@@ -875,21 +873,45 @@ def __init__(
...
+             query = {age: {"$gt: 5}}
+             MongoDBSource(connection_string=connection_string, collection_name="coll",
+             db_name="my_dataset", chunksize=5, query=query)
+
+        :param name:  source name
+        :param query: dictionary query for mongodb
+        :param chunksize: number of rows per chunk (default large single chunk)
+        :param key_field: the column to be used as the key for events. Can be a list of keys.
+        :param time_field: the column to be parsed as the timestamp for events. Defaults to None
+        :param start_time: filters out data before this time
```
Comment (guy1992l) [inline mlrun/datastore/sources.py:890]: why `start_time` and `end_time` do not have type hints? If they can be multiple types, you can use `typing.Union[int, str]` for exmaple to be either `str` or `int`.

File: mlrun/datastore/sources.py [hunk 13]
```diff
@@ -905,41 +927,32 @@ def __init__(
         )
 
     def to_dataframe(self):
```
Comment (guy1992l) [inline mlrun/datastore/sources.py:929]: it is a bit misleading that the method is called to_dataframe but an `MongoDBIterator` can be returned instead of a `pandas.DataFrame`... try to give a better name and add type hints for the returning value:

```
def to_dataframe(self) -> Union[pd.DataFrame, _MondoDBIterator]:
```

File: mlrun/datastore/sources.py [hunk 14]
```diff
@@ -52,6 +51,30 @@ def get_source_step(source, key_fields=None, time_field=None, context=None):
     return source.to_step(key_fields, time_field, context)
 
 
+class _MongoDBiter:
```
Comment (guy1992l) [inline mlrun/datastore/sources.py:54]: either `_MongoDBIter` or my preference: `_MongoDBIterator`. depends on the naming style you see in this file...

File: mlrun/datastore/targets.py [hunk 15]
```diff
@@ -1470,11 +1458,26 @@ def __init__(
...
+
+        :param db_name:             the name of the database to access
+        :param connection_string:   your mongodb connection string
+        :param collection_name:     the name of the collection to access,
+                                    from the current database
+        :param create_collection:   pass True if you want to create new collection named by
+                                    collection_name on current database.
+        :param override_collection: pass True if you want to override all the documents on the
+                                    collection named by collection_name on current database.
+        """
```
Comment (guy1992l) [inline mlrun/datastore/targets.py:1472]: superb docs 💪 !!! why can't you do these for all the ones you are writing 🥺 ?

File: mlrun/datastore/targets.py [hunk 16]
```diff
@@ -1432,18 +1432,6 @@ def as_df(
 
 
 class MongoDBTarget(BaseStoreTarget):
```
Comment (guy1992l) [inline mlrun/datastore/targets.py:1434]: maybe add an example like you did for source? or perhaps all the targets are examples-less?

File: mlrun/datastore/targets.py [hunk 17]
```diff
@@ -1496,13 +1500,15 @@ def __init__(
...
             elif override_collection:
+                # override all the collection's documents
                 db[collection_name].delete_many({})
             attr = {
                 "collection_name": collection_name,
                 "db_name": db_name,
                 "connection_string": connection_string,
             }
             path = f"mdb:///{connection_string}///{db_name}///{collection_name}"
+            mongodb_client.close()
```
Comment (guy1992l) [inline mlrun/datastore/targets.py:1511]: sweet 🤙 

But, I see you open and close multiple times, think if it's better to have the session open and closed once at the time the class is created and deleted (via overriding the `__del__` method). I'm not familliar with the logic of target so its just a note for you, not a thing to necessarily change.

File: mlrun/datastore/targets.py [hunk 18]
```diff
@@ -1593,6 +1599,7 @@ def as_df(
         except KeyError:
             query = {}
```
Comment (guy1992l) [inline mlrun/datastore/targets.py:1600]: much more healthy way is to do:
```
query = kwargs.get("query", {})
``` 
no need for a try-except code here.

**Review by guy1992l — CHANGES_REQUESTED** [review]
Please add docs and type hints my man

File: mlrun/datastore/targets.py (hunk 15 above)
Comment (davesh0812) [reply mlrun/datastore/targets.py:1472]: thank :)

File: mlrun/datastore/targets.py (hunk 17 above)
Comment (davesh0812) [reply mlrun/datastore/targets.py:1511]: I think it is better like this because every time I open a connection I closed it in the end of the same method. @guy1992l

File: mlrun/datastore/sources.py (hunk 13 above)
Comment (davesh0812) [reply mlrun/datastore/sources.py:929]: I cannot change the name because it part of the api.. @guy1992l, but I add the typing.

File: mlrun/datastore/sources.py [hunk 19]
```diff
@@ -866,8 +866,8 @@ def __init__(
         key_field: str = None,
         time_field: str = None,
         schedule: str = None,
-        start_time=None,
-        end_time=None,
+        start_time: Optional[Union[datetime, str]] = None,
```
Comment (guy1992l) [inline mlrun/datastore/sources.py:869]: the type hint Optional is mostly for `dataclass` objects so the methods will be set accordingly. When you have `= None` it is optional by default :)

File: mlrun/datastore/sources.py [hunk 20]
```diff
@@ -877,34 +877,36 @@ def __init__(
...
+        :param db_name:             the name of the database to access
+        :param connection_string:   your mongodb connection string
+        :param collection_name:     the name of the collection to access,
                                         from the current database
-        :param spark_options: additional spark read options
+        :param spark_options:       additional spark read options
         """
 
-        MONGO_CONNECTION_STRING_ENV_VAR = "MONGO_CONNECTION_STRING"
+        _MONGO_CONNECTION_STRING_ENV_VAR = "MONGO_CONNECTION_STRING"
```
Comment (guy1992l) [inline mlrun/datastore/sources.py:903]: Move to be a class variable:
```
class Guy:
    # A constant for the environment variable name of the bla bla:
    _MONGO_CONNECTION_STRING_ENV_VAR = "MONGO_CONNECTION_STRING"
```

File: mlrun/datastore/targets.py [hunk 21]
```diff
@@ -1594,10 +1600,7 @@ def as_df(
         time_column=None,
         **kwargs,
     ):
-        try:
-            query = kwargs["query"]
```
Comment (guy1992l) [inline mlrun/datastore/targets.py]: you don't need the query in `kwargs` anymore? You could have done `query = kwargs.get("query", {})`

**Review by guy1992l — CHANGES_REQUESTED** [review]

File: mlrun/datastore/targets.py (hunk 17 above)
Comment (guy1992l) [reply mlrun/datastore/targets.py:1517]: consult with @gtopper on this one and let's hear his opinion as well :) Or maybe in the pymongo examples this is the way they recommend to work?

Comment (davesh0812) [reply mlrun/datastore/targets.py:1517]: @guy1992l they recommended to close after each use, and also we have few classes that can read from the same mongo data base, so it is safer to close when it possible.
